backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java

```diff
@@ -2251,52 +2262,173 @@ public void deleteLogVersionedInputs(Session session, Long repoId, String commit
-          }
-          return new AbstractMap.SimpleEntry<>(artifactEntity.getPath(), uploadId);
-        }
+      if (!Objects.equals(uploadId, artifactEntity.getUploadId())
+          || artifactEntity.isUploadCompleted()) {
+        session.beginTransaction();
+        artifactEntity.setUploadId(uploadId);
+        artifactEntity.setUploadCompleted(false);
+        session.getTransaction().commit();
       }
-      return new AbstractMap.SimpleEntry<>(null, null);
+    } else {
+      uploadId = null;
+    }
+    return new AbstractMap.SimpleEntry<>(artifactEntity.getPath(), uploadId);
+  }
+
+  @Override
+  public Response commitArtifactPart(CommitArtifactPart request)
+      throws ModelDBException, InvalidProtocolBufferException {
+    try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
+      ArtifactEntity artifactEntity =
+          getArtifactEntity(
+              session, request.getId(), request.getKey(), ModelDBServiceActions.UPDATE);
+      ArtifactPart artifactPart = request.getArtifactPart();
+      ArtifactPartEntity artifactPartEntity =
+          new ArtifactPartEntity(
+              artifactEntity, artifactPart.getPartNumber(), artifactPart.getEtag());
+      session.beginTransaction();
+      session.saveOrUpdate(artifactPartEntity);
+      session.getTransaction().commit();
+      return Response.newBuilder().build();
+    }
+  }
+
+  @Override
+  public GetCommittedArtifactParts.Response getCommittedArtifactParts(
+      GetCommittedArtifactParts request) throws ModelDBException, InvalidProtocolBufferException {
+    try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
+      Set<ArtifactPartEntity> artifactPartEntities =
+          getArtifactPartEntities(session, request.getId(), request.getKey());
+      GetCommittedArtifactParts.Response.Builder response =
+          GetCommittedArtifactParts.Response.newBuilder();
+      artifactPartEntities.forEach(
+          artifactPartEntity -> response.addArtifactParts(artifactPartEntity.toProto()));
+      return response.build();
+    }
+  }
+
+  private Set<ArtifactPartEntity> getArtifactPartEntities(
+      Session session, String experimentRunId, String key)
+      throws ModelDBException, InvalidProtocolBufferException {
+    ArtifactEntity artifactEntity =
+        getArtifactEntity(session, experimentRunId, key, ModelDBServiceActions.READ);
+    return artifactEntity.getArtifactPartEntities();
+  }
+
+  @Override
+  public CommitMultipartArtifact.Response commitMultipartArtifact(
+      CommitMultipartArtifact request, CommitMultipartFunction commitMultipartFunction)
+      throws ModelDBException, InvalidProtocolBufferException {
+    List<PartETag> partETags;
+    try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
+      ArtifactEntity artifactEntity =
+          getArtifactEntity(
+              session, request.getId(), request.getKey(), ModelDBServiceActions.UPDATE);
```

Comment:
This will involve a grpc to check permission. You should not do that while holding a db session.

Reply:
https://github.com/VertaAI/modeldb/pull/691